### PR TITLE
identity: adds generation of plugin identity tokens

### PR DESCRIPTION
**Pull request by austingebauer**
This PR introduces generation of plugin identity tokens to the identity store. This allows plugins to consume them over the `GenerateIdentityToken()` method of the system view.

Other important aspects of this PR:
- Prevents [deletion of keys](https://developer.hashicorp.com/vault/api-docs/secret/identity/tokens#delete-a-named-key) that are in use by a plugin mount
- Lazily generates the OIDC default key if used at plugin mount time

I've added a test that generates JWTs, verifies their signatures, and validates a majority of claims. A test that integrates with a plugin consuming these tokens (AWS secret engine) will be added in a follow-up PR.

### CODE REVIEW

File: vault/identity_store_oidc.go [hunk 1]
```diff
@@ -1028,6 +1095,91 @@ func (i *IdentityStore) pathOIDCGenerateToken(ctx context.Context, req *logical.
...
+		"iss": issuer,
+		"sub": fmt.Sprintf("%s:%s:%s:%s", pluginTokenSubjectPrefix, nsID, me.Table, me.Accessor),
+		"aud": []string{audience},
+		"nbf": now.Unix(),
+		"iat": now.Unix(),
+		"exp": now.Add(ttl).Unix(),
+		pluginTokenPrivateClaimKey: map[string]any{
+			"namespace_id":   ns.ID,
+			"namespace_path": ns.Path,
+			"class":          me.Table,
```
Comment (tomhjp) [inline vault/identity_store_oidc.go:1162]: `me.Table` will either be `mounts` for secrets plugins, or `auth` for auth plugins, so I think we might need to do a little bit of translation for secrets plugins. I think it should probably be `secret` (i.e. not the plural form) to match `auth`.

File: vault/logical_system.go [hunk 2]
```diff
@@ -1521,6 +1522,15 @@ func (b *SystemBackend) handleMount(ctx context.Context, req *logical.Request, d
 		config.IdentityTokenKey = apiConfig.IdentityTokenKey
 	}
 
+	// Don't lazily generate the default OIDC key for KV mounts. A default KV mount
```
Comment (tomhjp) [inline vault/logical_system.go:1525]: TIL, I wouldn't have thought the HTTP mount handler function would be in the call path for the default KV mount.

**Review by tomhjp — APPROVED** [review]
LGTM! 👍 just a few small suggestions

File: changelog/25219.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+```release-note:feature
+**Plugin Identity Tokens**: Gives plugins a source of identity that can be used for secret-less configuring using workload identity federation.
```
[the remark this answers is not in the document]
Comment (austingebauer) [reply changelog/25219.txt]: Ah yes, we should. I like that better. Thanks 👍

File: vault/identity_store_oidc.go [hunk 4]
```diff
@@ -1028,6 +1095,91 @@ func (i *IdentityStore) pathOIDCGenerateToken(ctx context.Context, req *logical.
...
+	}
+
+	// Tokens for plugins have a distinct issuer from Vault's identity token issuer
+	issuer, err := config.fullIssuer(pluginIdentityTokenIssuer)
+	if err != nil {
+		return "", 0, err
+	}
+
+	var nsID string
+	if me.Namespace() != nil {
```
[the remark this answers is not in the document]
Comment (fairclothjm) [reply vault/identity_store_oidc.go:1147]: 👍

**Review by fairclothjm — APPROVED** [review]
+1 to Tom's suggestions, otherwise LGTM!

File: vault/identity_store_oidc.go (hunk 1 above)
Comment (austingebauer) [reply vault/identity_store_oidc.go:1162]: Cool, this was on my mind at some point. I was thinking the same (change to singular `secret`). I'll make that change!

File: vault/identity_store_oidc.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (austingebauer) [reply vault/identity_store_oidc.go]: Ah, right. This came from some testing where I didn't populate the namespace on a mount entry. I removed it and am using the namespace from the context instead in https://github.com/hashicorp/vault/pull/25219/commits/b848fcbb161352dc98074083e14a5504a9fc474f. Thanks!

File: vault/identity_store_oidc.go (hunk 1 above)
Comment (austingebauer) [reply vault/identity_store_oidc.go]: Done in https://github.com/hashicorp/vault/pull/25219/commits/f03ac93a284aba995d2ae11d02ad4ef2fb55a8d9. Thanks!

**Pull request merged**